src/Week10/AParser.hs

```diff
@@ -0,0 +1,172 @@
+    
+{-
+Exercise 2
+Now implement an Applicative instance for Parser:
+• pure a represents the parser which consumes no input and suc- cessfully returns a result of a.
+• p1 <*> p2 represents the parser which first runs p1 (which will consume some input and produce a function), 
+then passes the remaining input to p2 (which consumes more input and produces some value), 
+then returns the result of applying the function to the value. 
+However, if either p1 or p2 fails then the whole thing should also fail (put another way, p1 <*> p2 only succeeds if both p1 and p2 succeed).
+So what is this good for? Recalling the Employee example from class,
+type Name = String
+ data Employee = Emp { name :: Name, phone :: String }
+we could now use the Applicative instance for Parser to make an employee parser from name and phone parsers. 
+That is, if
+ parseName  :: Parser Name
+ parsePhone :: Parser String
+then
+ Emp <$> parseName <*> parsePhone :: Parser Employee
+is a parser which first reads a name from the input, then a phone number, and returns them combined into an Employee record. 
+Of course, this assumes that the name and phone number are right next to each other in the input, with no intervening separators. 
+We’ll see later how to make parsers that can throw away extra stuff that doesn’t directly correspond to information you want to parse.
+-}
+
+instance Applicative Parser where
+    pure a = Parser $ (\s -> Just (a, s))
+    p1 <*> p2 = Parser p
+        where 
+            p s = runParser p1 s >>= g
+            g (f, res_str) = runParser (f <$> p2) res_str
+
+{-
+Exercise 3
+We can also test your Applicative instance using other simple
+applications of functions to multiple parsers. 
+You should implement each of the following exercises using the Applicative interface to put together simpler parsers into more complex ones. 
+Do not implement them using the low-level definition of a Parser! In other words, pre- tend that you do not have access to the Parser constructor or even know how the Parser type is defined.
+• Create a parser
+abParser :: Parser (Char, Char)
+which expects to see the characters ’a’ and ’b’ and returns them
+as a pair. That is,
+   *AParser> runParser abParser "abcdef"
+   Just ((’a’,’b’),"cdef")
+   *AParser> runParser abParser "aebcdf"
+   Nothing
+• Now create a parser
+abParser_ :: Parser ()
+which acts in the same way as abParser but returns () instead of the characters ’a’ and ’b’.
+*AParser> runParser abParser_ "abcdef" Just ((),"cdef")
+*AParser> runParser abParser_ "aebcdf" Nothing
+• Create a parser intPair which reads two integer values separated by a space and returns the integer values in a list.
+ You should use the provided posInt to parse the integer values.
+   *Parser> runParser intPair "12 34"
+   Just ([12,34],"")
+-}
+
+abParser :: Parser (Char, Char)
+abParser = (,) <$> char 'a' <*> char 'b'
+
+abParser_ :: Parser ()
+abParser_ =  const () <$> abParser
```

Comment:
You can write the function like this as well:
```
abParser_ = () <$ abParser
```
The operator `<$` can be interpreted as `fmap . const`. You might want to check it out.